src/lib/backoff.ts

```diff
@@ -0,0 +1,37 @@
+import { log } from '../lib/supervisor-console';
+import sleep from '../lib/sleep';
+
+/**
+ * Perform exponential backoff on the function, increasing the attempts
+ * counter on each call
+ *
+ * If attempts is 0 then, it will delay for min(minDelay, maxDelay)
+ */
+export const backoff = (
+	retry: (attempts: number) => void,
+	attempts = 0,
+	maxDelay: number,
+	minDelay: number,
+) => {
+	const delay = Math.min(2 ** attempts * minDelay, maxDelay);
+	log.info(`Retrying in ${delay / 1000} seconds`);
+	setTimeout(() => retry(attempts + 1), delay);
+};
+
+/**
+ * Perform exponential backoff on an async function, increasing the attempts
+ * counter on each call
+ *
+ * If attempts is 0 then, it will delay for min(minDelay, maxDelay)
+ */
+export async function backoffPromise(
```

Comment:
I don't see the point of having both functions, they both do exactly the same thing. We should go with the promise version which is the right way we want to do this type of operation, what do you think?

Reply:
I agree, if we did want to keep both versions we could just do:
```ts
export const backoff = (
	...args: Parameters<typeof backoffPromise>
) => {
	backoffPromise(...args);
};
```
and have the same result with less code